rfc/hierarchical-capabilities.md

```diff
@@ -0,0 +1,57 @@
+# Hierarchical Capabilities
+
+## Abstract
+
+Currently UCAN capabilities described in [W3Up specification](https://github.com/web3-storage/specs) do not follow consistent organizational structure. To make things worse accounts on login delegate `*` capabilities and spaces delegate set of namespaces.
+
+This creates a serious challenge for the programs (like w3up client) that attempt to infer set of methods to expose based no delegated capabilities.
+
+We roughly two roles in the sysntem `space` and `account` and their capability set do not overlap, so we can somewhat infer which methods to present in case of `*` ability by subject DID, but that hinges upon the fact that we use `did:key` for space and `did:mailto` for accounts. This assumbtion may be rendered obsolete in the future.
+
+## Proposal
+
+Re-organize capabilities into role based namespeces e.g. `/space/` and `/account/` making it possible for the program to infer which set of methods to expose. All of the `store/*` and `upload/*` etc capabilities SHOULD be remapped under `/space/` namespace.
+
+In addition we should capture depndency across namespaces in the way capabilities are orginalizated e.g. if upload task requires storing bytes we nest `store` capabilities under `upload` here is an example layout that may work
+
+
+```sh
+# write permission
+/space/content/add                          # upload/add
+/space/content/add/shard                    # store/add
+/space/content/add/shard/allocate           # store/allocate
+/space/content/add/shard/write              # S3/R2 PUT
```

Comment:
I am not sure about this proposal in specific. I would not consider `upload/add` to be a "parent" of `store/add` in a way. Probably I would put it into a different layer, like `space/collection` and drop `shard` from here instead?

Reply:
Yes, something more along the lines of `/space/content/add/root`, `/space/content/add/pack` I think would feel better

Reply:
> I also wonder if this is the correct order - I was thinking perhaps the verb should be at the end? Should the noun(s) drill down to the "action" (verb) from `/`? So you can read the capability right to left and it make sense?

While I do share the sentiment problem is that when user wants to share ability to add content they would have to delegate several capabilities as you need be able to add shards and uploads etc...

Which is also why `content/add` contained all kinds of writes so user can just delegate white capability, or limit it to specific write capability where needed.

Reply:
> It feels more likely to me that users would want to "wildcard-delegate" based on the thing (noun) not the action (verb)....probably not explaining that well, but I'd be hesitant to delegate /space/add/* - ability to add anything that exists or will exist in the future (narrow action, wide things), whereas /space/blob/* - any action with blobs specifically feels safer, despite widening of granted actions (narrow thing, wide actions).

Well that is why I had `/space/content/` to cover everything content related wheres `/space/content/add` covered all content writes and if you want to get very specific that `/space/content/add/blob/`.

So if we replace `blob` with `content` you get basically what you're suggesting here. However I wanted `content` to cover blobs (which are effectively shards) and DAGs consisting of blocks from shards.

Perhaps `content` is not best term for it ?

Reply:
I still think content/blob would be important because we want to have caps to add side indexes and so one to the space? While they coul in theory be a content themselves, I think making distinction would be good so that we can treat those more as hidden files in console